src/mainboard/pcengines/apu2/mainboard.c

```diff
@@ -284,4 +290,51 @@ struct chip_operations mainboard_ops = {
 	.final = mainboard_final,
 };
 
+const char *smbios_mainboard_serial_number(void)
+{
+    static char serial[10];
+    msr_t msr;
+    u32 mac_addr = 0;
+    device_t nic_dev;
+
+    // Allows the IO configuration space access method, IOCF8 and IOCFC, to be
+    // used to generate extended configuration cycles
+    msr = rdmsr(NB_CFG_MSR);
+    msr.hi |= (ENABLE_CF8_EXT_CFG);
+    wrmsr(NB_CFG_MSR, msr);
+
+    nic_dev = dev_find_slot(1, PCI_DEVFN(0, 0));
+
+    if ((serial[0] != 0) || !nic_dev)
+        return serial;
+
+    // Read three LSB of MAC address
+    mac_addr = pci_read_config32(nic_dev, 0x144);
```

Comment:
@macpijan `0x144`. Maybe there is define for that or it is worth to call this address as described by tools like `lspci`.